File: ts/test/chat/privacy.ts
import { Filter } from "firebase-admin/firestore";
import { FirestoreLocator, Locator, LocatorType } from "../../src/model";
import GroupChat from "./model/gc";
import Message from "./model/message";
import { FirestoreCollections } from "./model/shared";
import User from "./model/user";

export default function handleAccess(dataSubjectId: string, locator: FirestoreLocator, obj: any): Record<string, any> {
    switch (locator.dataType) {
        case 'user':
            return handleAccessUser(dataSubjectId, locator, obj as User);
        case 'groupChat':
            return handleAccessGroupChat(dataSubjectId, locator, obj as GroupChat);
        case 'message':
            return handleAccessMessage(dataSubjectId, locator, obj as Message);
        default:
            throw new Error(`Data type ${locator.dataType} does not have handleAccess function implemented`);
    }
}

function handleAccessGroupChat(dataSubjectId: string, locator: FirestoreLocator, obj: GroupChat): Record<string, any> {
    return {
        messages: {
            dataType: 'message',
            singleDocument: false,
            locatorType: LocatorType.Collection,
            collectionPath: [...locator.collectionPath, FirestoreCollections.Messages],
            docIds: locator.docIds,
            queries: [Filter.where('userID', '==', dataSubjectId)]
        } as FirestoreLocator
    };
};

function handleAccessUser(dataSubjectId: string, locator: FirestoreLocator, obj: User): Record<string, any> {
    return {
        name: obj.name,
        groupChats: obj.gcs.map((gc: string): FirestoreLocator => {
            return {
                dataType: 'groupChat',
                singleDocument: true,
                locatorType: LocatorType.Document,
                collectionPath: [FirestoreCollections.GroupChat],
                docIds: [gc],
            }
        })
    };
}

function handleAccessMessage(dataSubjectId: string, locator: Locator, obj: Message): Record<string, any> {
    return {
        content: obj.content,
        timestamp: obj.timestamp
    };
};

File: go/internal/test/chat/privacy.go
package chat

import (
	pal "github.com/privacy-pal/privacy-pal/pkg"
)

const (
	UserDataType          = "user"
	GroupChatDataType     = "groupchat"
	DirectMessageDataType = "directmessage"
	MessageDataType       = "message"
)

func HandleAccess(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	switch currentDbObjLocator.DataType {
	case UserDataType:
		return HandleAccessUser(dataSubjectId, currentDbObjLocator, dbObj)
	case GroupChatDataType:
		return HandleAccessGroupChat(dataSubjectId, currentDbObjLocator, dbObj)
	case MessageDataType:
		return HandleAccessMessage(dataSubjectId, currentDbObjLocator, dbObj)
	case DirectMessageDataType:
		return HandleAccessDirectMessage(dataSubjectId, currentDbObjLocator, dbObj)
	default:
		// TODO: should return error
		return nil
	}
}

func HandleAccessUser(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	data := make(map[string]interface{})

	// TODO: include in documentation: you can access the id in 2 ways
	if dbObj["_id"].(string) != dataSubjectId {
		data["Name"] = dbObj["name"]
		return data
	}
	// if currentDbObjLocator.DocIDs[len(currentDbObjLocator.DocIDs)-1] != dataSubjectId {
	// 	data["Name"] = dbObj["name"]
	// 	return data
	// }

	data["Name"] = dbObj["name"]
	data["Groupchats"] = make([]pal.Locator, 0)
	for _, id := range dbObj["gcs"].([]interface{}) {
		id := id.(string)
		data["Groupchats"] = append(data["Groupchats"].([]pal.Locator), pal.Locator{
			LocatorType: pal.Document,
			DataType:    GroupChatDataType,
			FirestoreLocator: pal.FirestoreLocator{
				CollectionPath: []string{"gcs"},
				DocIDs:         []string{id},
			},
		})
	}
	data["DirectMessages"] = make([]pal.Locator, 0)
	// TODO: support this in yaml
	for _, DMId := range dbObj["dms"].(map[string]interface{}) {
		DMId := DMId.(string)
		data["DirectMessages"] = append(data["DirectMessages"].([]pal.Locator), pal.Locator{
			LocatorType: pal.Document,
			DataType:    DirectMessageDataType,
			FirestoreLocator: pal.FirestoreLocator{
				CollectionPath: []string{"dms"},
				DocIDs:         []string{DMId},
			},
		})
	}

	return data
}

func HandleAccessGroupChat(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	data := make(map[string]interface{})

	data["Messages"] = pal.Locator{
		LocatorType: pal.Collection,
		DataType:    MessageDataType,
		FirestoreLocator: pal.FirestoreLocator{
			CollectionPath: append(currentDbObjLocator.FirestoreLocator.CollectionPath, "messages"),
			DocIDs:         currentDbObjLocator.DocIDs,
			Queries: []pal.Query{
				{
					Path:  "userId",
					Op:    "==",
					Value: dataSubjectId,
				},
			},
		},
	}

	return data
}

func HandleAccessMessage(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	data := make(map[string]interface{})

	data["Content"] = dbObj["content"]
	data["Timestamp"] = dbObj["timestamp"]

	return data
}

func HandleAccessDirectMessage(dataSubjectId string, currentDbObjLocator pal.Locator, dbObj pal.DatabaseObject) map[string]interface{} {
	data := make(map[string]interface{})

	var otherUserId string
	if dbObj["user1"].(string) == dataSubjectId {
		otherUserId = dbObj["user2"].(string)
	} else {
		otherUserId = dbObj["user1"].(string)
	}

	data["Other User"] = pal.Locator{
		LocatorType: pal.Document,
		DataType:    UserDataType,
		FirestoreLocator: pal.FirestoreLocator{
			CollectionPath: []string{"users"},
			DocIDs:         []string{otherUserId},
		},
	}
	data["Messages"] = pal.Locator{
		LocatorType: pal.Collection,
		DataType:    MessageDataType,
		FirestoreLocator: pal.FirestoreLocator{
			CollectionPath: append(currentDbObjLocator.FirestoreLocator.CollectionPath, "messages"),
			DocIDs:         currentDbObjLocator.DocIDs,
			Queries: []pal.Query{
				{
					Path:  "userId",
					Op:    "==",
					Value: dataSubjectId,
				},
			},
		},
	}

	return data
}
